### ISSUE TITLE
Error: Missing template processor

### ISSUE THREAD

**Issue by elwayman02** [ember-cli/ember-cli#3419]
Created a fresh Ember addon last night and made minor changes.  Running ```npm test``` on the addon or consuming the addon in an app and running ```ember s``` throws the following error:

```
> ember test

version: 0.2.0-beta.1
Could not find watchman, falling back to NodeWatcher for file system events
Building...
Missing template processor
Error: Missing template processor
    at module.exports.preprocessTemplates (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/preprocessors.js:149:11)
    at Class.Addon.compileTemplates (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/models/addon.js:434:12)
    at Class.Addon.compileAddon (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/models/addon.js:456:28)
    at Class.Addon.treeForAddon (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/models/addon.js:382:24)
    at Class._treeFor (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/models/addon.js:282:31)
    at Class.treeFor (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/models/addon.js:253:19)
    at EmberAddon.<anonymous> (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/broccoli/ember-app.js:307:20)
    at Array.map (native)
    at EmberAddon.EmberApp.addonTreesFor (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/broccoli/ember-app.js:305:30)
    at EmberAddon._addonTree (/Users/hawkerj/Projects/ember-bug-widget/node_modules/ember-cli/lib/broccoli/ember-app.js:597:38)
npm ERR! Test failed.  See above for more details.
```

Ember-cli 0.2.0-beta.1
node 0.12.0
npm 2.1.8

Here's a diff of the changes I made: https://github.com/elwayman02/ember-bug-widget/commit/01af8d5ca9f18c16a1388c1711f9b09f190785a5

To reproduce:

1) Checkout ember-bug-widget
2) Run ```npm install```
3) Run ```npm test```

or

1) Consume ember-bug-widget as an addon in an Ember app
2) Run ```ember s```

**Comment by chadhietala** [comment]
Can you preform the following:

```
rm -rf node_modules && npm cache clean && npm install
```

**Comment by elwayman02** [comment]
@chadhietala I get the same error

**Comment by rwjblue** [comment]
If your addon has a addon/templates/ folder, you will need to ensure that ember-cli-htmlbars is listed in your addon's package.json as a dependency (not a devDependency).  A much better error message was added in a PR a few days ago, and will be included in 0.2.0 final. 

Closing as I think the above should fix the issue, but happy to reopen with steps to reproduce otherwise.

**Comment by bcardarella** [comment]
@rwjblue why are we not including the `ember-cli-htmlbars` as a development dependency for addons by default? It seems like this is going to be a very annoying troll for people

**Comment by rwjblue** [comment]
It is a dev dep by default. You are asking why is it not a dep. 

It is not a dep by default because it is only needed if your addon has templates in addon/templates (which only applies to a subset of addons). 

As I mentioned above we added a much more useful error message in master when this misalignmet occurs. We should make sure that the error is thrown when using the addons dummy app (and therefore failing the build and tests of the addon itself).

**Comment by elwayman02** [comment]
@rwjblue maybe this is something that could be mentioned in the README that gets auto-generated when a new addon is created?  It seems like having templates in an addon will be a pretty common use-case

**Comment by elwayman02** [comment]
@rwjblue moving the dependency caused a new error: 

```
> ember-bug-widget@0.0.1 test c:\Projects\GitHub\ember-bug-widget
> ember test

version: 0.2.0-beta.1
Built project successfully. Stored in "c:\Projects\GitHub\ember-bug-widget\tmp\c
lass-tests_dist-sGjWeB9R.tmp".

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: spawn ENOENT
    at errnoException (child_process.js:1001:11)
    at Process.ChildProcess._handle.onexit (child_process.js:792:34)
npm ERR! Test failed.  See above for more details.
npm ERR! not ok code 0
```

Consuming the addon into an app with this change still shows the "Missing template processor" error.

**Comment by stefanpenner** [comment]
> It is not a dep by default because it is only needed if your addon has templates in addon/templates (which only applies to a subset of addons).

i think if someone generates something that needs a template, we should add it to deps

**Comment by rwjblue** [comment]
@stefanpenner - agreed. Also, we should ensure (aka add some tests) that 'ember test' fails in a generated addon when addon/templates exists but a template compiler is not in deps.

**Comment by stefanpenner** [comment]
> @rwjblue moving the dependency caused a new error:

**Comment by rwjblue** [comment]
I cloned the referenced repo, and do not receive an error on `ember build` or `ember test`...

**Comment by stefanpenner** [comment]
@elwayman02 can you share:

* ember-cli version
* node version
* npm version
* windows version

**Comment by elwayman02** [comment]
Logged #3480 for this new error, since it doesn't seem specific to the template issue.  Let's hold off on trying to solve this until we fix the npm test problem, since it's currently hiding the previous error.

My env:

Windows 7 Home
ember-cli 0.2.0-beta.1
node 0.12.0
npm 2.1.8

**Comment by elwayman02** [comment]
Upgrading to 0.2.0 to try again

**Comment by rwjblue** [comment]
@elwayman02 - The initially reported error should be fixed with my initial suggestion, but now latest error (tracking in #3480) needs to be addressed.

**Comment by elwayman02** [comment]
:+1: I look forward to being able to confirm this once #3480 is resolved :)

**Comment by simo163** [comment]
I am having almost the same error, any suggestions?
```
mgharbi$ ember test
version: 0.2.0
Building...
Missing template processor
Error: Missing template processor
    at module.exports.preprocessTemplates (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/preprocessors.js:164:11)
    at EmberApp._processedTemplatesTree (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:532:10)
    at EmberApp.appAndDependencies (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:773:24)
    at EmberApp.javascript (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:838:34)
    at EmberApp.toArray (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:1153:10)
    at EmberApp.toTree (/Users/mgharbi/.nvm/versions/node/v0.12.0/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:1175:30)
    at Object.<anonymous> (/Users/mgharbi/ember_workspace/borrowers/Brocfile.js:20:22)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
```

**Comment by jimijon** [comment]
Just started... then tried to launch my new app:

$ ember server
version: 0.2.7
Could not find watchman, falling back to NodeWatcher for file system events.
Visit http://www.ember-cli.com/#watchman for more info.
Missing template processor
Error: Missing template processor
    at module.exports.preprocessTemplates (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/preprocessors.js:164:11)
    at EmberApp._processedTemplatesTree (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/broccoli/ember-app.js:567:10)
    at EmberApp.appAndDependencies (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/broccoli/ember-app.js:808:24)
    at EmberApp.javascript (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/broccoli/ember-app.js:882:34)
    at EmberApp.toArray (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/broccoli/ember-app.js:1250:10)
    at EmberApp.toTree (/Users/James/Documents/Development/scpq-ember/node_modules/ember-cli/lib/broccoli/ember-app.js:1272:30)
    at Object.<anonymous> (/Users/James/Documents/Development/scpq-ember/Brocfile.js:20:22)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)

**Comment by brianally** [comment]
Slightly different error for me, plus extra strangeness:

    $ ember new trender
    version: 0.2.7
    1.13.1

    installing
      create .bowerrc
      create ...
      ... etc
    ENOENT, open '/home/bally/apps/package.json'
    Error: ENOENT, open '/home/bally/apps/package.json'
        at Error (native)

    $ cd trender && ember server
    version: 0.2.7

    A new version of ember-cli is available (1.13.1). To install it, type ember update.
    Missing template processor
    Error: Missing template processor
        at module.exports.preprocessTemplates (/usr/lib/node_modules/ember-cli/lib/preprocessors.js:164:11)
        at EmberApp._processedTemplatesTree (/usr/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:532:10)
        at EmberApp.appAndDependencies (/usr/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:773:24)
        at EmberApp.javascript (/usr/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:838:34)
        at EmberApp.toArray (/usr/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:1153:10)
        at EmberApp.toTree (/usr/lib/node_modules/ember-cli/lib/broccoli/ember-app.js:1175:30)
        at Object.<anonymous> (/home/bally/apps/trender/Brocfile.js:20:22)
        at Module._compile (module.js:460:26)
        at Object.Module._extensions..js (module.js:478:10)
        at Module.load (module.js:355:32)


There are several oddities here that may, or may not, be relevant. When the new app is created it shows that i'm running 1.13.1, but when i attempt to run the server i'm notified that a new version is available -- but it's the same version.

Next, why the complaint that package.json cannot be found outside of the new app directory?

And then,of course, the missing template processor.

    $ ember --version
    version: 0.2.7

    A new version of ember-cli is available (1.13.1). To install it, type ember update.
    node: 0.12.0
    npm: 2.11.0

**Comment by stefanpenner** [comment]
Sounds like you need to run npm install

**Comment by technomage** [comment]
jimijon: I resolved this by doing npm install ember-cli-htmlbars --save and then it produced an error that indicated I needed to do ember install ember-cli-htmlbars-inline-precompile which was already in my package.json.  But, doing those 2 steps resolved the issue.  This in ember-cli 2.7 and ember 2.7 on a newly generated add on.  Note that I am using ember-cli-mocha which may have factored into it.
